lib/mcp/tool/input_schema.rb

```
@@ -21,6 +26,32 @@ def missing_required_arguments?(arguments)
       def missing_required_arguments(arguments)
         (required - arguments.keys.map(&:to_sym))
       end
+
+      def validate_arguments(arguments)
+        errors = JSON::Validator.fully_validate(to_h, arguments)
+        if errors.any?
+          raise ValidationError, "Invalid arguments: #{errors.join(", ")}"
+        end
+      end
+
+      private
+
+      def validate_schema!
+        schema = to_h
+        begin
+          schema_reader = JSON::Schema::Reader.new(
+            accept_uri: false,
+            accept_file: ->(path) { path.to_s.start_with?(Gem.loaded_specs["json-schema"].full_gem_path) },
+          )
+          metaschema = JSON::Validator.validator_for_name("draft4").metaschema
+          errors = JSON::Validator.fully_validate(metaschema, schema, schema_reader: schema_reader)
+          if errors.any?
+            raise ArgumentError, "Invalid JSON Schema: #{errors.join(", ")}"
+          end
+        rescue StandardError => e
+          raise ArgumentError, "Invalid JSON Schema: #{e.message}"
```

This exception handling looked a bit odd to me, as it uses a generic exception class like `StandardError` to represent a specific error such as "Invalid JSON Schema". It might be more appropriate either to let the exception propagate to the caller without handling it here, or to replace `StandardError` with a more specific exception class.

---

Good point.